diff --git a/src/client/utils/settings.cpp b/src/client/utils/settings.cpp
index 380794fd..ccb75cc2 100644
--- a/src/client/utils/settings.cpp
+++ b/src/client/utils/settings.cpp
@@ -42,6 +42,7 @@ QString recordingFolder()
 QStringList dataPaths()
 {
 	QStringList datapaths;
+	datapaths << "share/drawpile/drawpile";
 #ifndef Q_OS_MAC
 	datapaths << qApp->applicationDirPath();
 	datapaths << QStandardPaths::standardLocations(QStandardPaths::DataLocation);
diff --git a/src/shared/CMakeLists.txt b/src/shared/CMakeLists.txt
index 1cd36110..cb4508db 100644
--- a/src/shared/CMakeLists.txt
+++ b/src/shared/CMakeLists.txt
@@ -1,6 +1,5 @@
 find_package(Qt5Network REQUIRED)
 find_package(KF5Archive REQUIRED NO_MODULE)
-find_package(Sodium)
 
 set (
 	SOURCES
