--- a/src/widgets/styles/qstylesheetstyle.cpp	2023-04-30 12:14:01.061389033 +0200
+++ b/src/widgets/styles/qstylesheetstyle.cpp	2023-04-30 12:14:09.778056012 +0200
@@ -1465,7 +1465,14 @@
         p->setBrush(cg, w->foregroundRole(), pal->foreground);
         p->setBrush(cg, QPalette::WindowText, pal->foreground);
         p->setBrush(cg, QPalette::Text, pal->foreground);
-        p->setBrush(cg, QPalette::PlaceholderText, pal->foreground);
+        // Drawpile patch: fix QTBUG-92199 that makes placeholder text be fully
+        // bright. This got fixed in 5.15.12, the code is from here:
+        // https://codereview.qt-project.org/c/qt/qtbase/+/434310
+        QColor phColor(pal->foreground.color());
+        phColor.setAlpha((phColor.alpha() + 1) / 2);
+        QBrush placeholder = pal->foreground;
+        placeholder.setColor(phColor);
+        p->setBrush(cg, QPalette::PlaceholderText, placeholder);
     }
     if (pal->selectionBackground.style() != Qt::NoBrush)
         p->setBrush(cg, QPalette::Highlight, pal->selectionBackground);
