diff -ur a/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qendian.h b/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qendian.h
--- a/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qendian.h	2020-10-27 09:02:11.000000000 +0100
+++ b/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qendian.h	2023-03-03 04:25:35.920508502 +0100
@@ -47,6 +47,7 @@
 // include stdlib.h and hope that it defines __GLIBC__ for glibc-based systems
 #include <stdlib.h>
 #include <string.h>
+#include <limits>
 
 #ifdef min // MSVC
 #undef min
diff -ur a/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qfloat16.h b/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qfloat16.h
--- a/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qfloat16.h	2020-10-27 09:02:11.000000000 +0100
+++ b/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qfloat16.h	2023-03-03 04:25:40.810508483 +0100
@@ -44,6 +44,7 @@
 #include <QtCore/qglobal.h>
 #include <QtCore/qmetatype.h>
 #include <string.h>
+#include <limits>
 
 #if defined(QT_COMPILER_SUPPORTS_F16C) && defined(__AVX2__) && !defined(__F16C__)
 // All processors that support AVX2 do support F16C too. That doesn't mean
diff -ur a/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qlogging.cpp b/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qlogging.cpp
--- a/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qlogging.cpp	2020-10-27 09:02:11.000000000 +0100
+++ b/qt-everywhere-src-5.15.2/qtbase/src/corelib/global/qlogging.cpp	2023-03-03 04:53:44.907166671 +0100
@@ -101,16 +101,6 @@
 #include <emscripten/emscripten.h>
 #endif
 
-#if QT_CONFIG(regularexpression)
-#  ifdef __UCLIBC__
-#    if __UCLIBC_HAS_BACKTRACE__
-#      define QLOGGING_HAVE_BACKTRACE
-#    endif
-#  elif (defined(__GLIBC__) && defined(__GLIBCXX__)) || (__has_include(<cxxabi.h>) && __has_include(<execinfo.h>))
-#    define QLOGGING_HAVE_BACKTRACE
-#  endif
-#endif
-
 #if QT_CONFIG(slog2)
 extern char *__progname;
 #endif
diff -ur a/qt-everywhere-src-5.15.2/qtbase/src/corelib/text/qbytearraymatcher.h b/qt-everywhere-src-5.15.2/qtbase/src/corelib/text/qbytearraymatcher.h
--- a/qt-everywhere-src-5.15.2/qtbase/src/corelib/text/qbytearraymatcher.h	2020-10-27 09:02:11.000000000 +0100
+++ b/qt-everywhere-src-5.15.2/qtbase/src/corelib/text/qbytearraymatcher.h	2023-03-03 04:26:35.097174883 +0100
@@ -41,6 +41,7 @@
 #define QBYTEARRAYMATCHER_H
 
 #include <QtCore/qbytearray.h>
+#include <limits>
 
 QT_BEGIN_NAMESPACE
 
