From 1fb8365368c285c0b9685442edc5eb9299bd6c25 Mon Sep 17 00:00:00 2001
From: drawdance <drawdance@example.com>
Date: Sat, 18 Sep 2021 02:24:51 +0200
Subject: [PATCH] Drawdance patches

---
 backends/imgui_impl_opengl3.cpp | 8 +-------
 1 file changed, 1 insertion(+), 7 deletions(-)

diff --git a/backends/imgui_impl_opengl3.cpp b/backends/imgui_impl_opengl3.cpp
index 6363dacd..11ecd837 100644
--- a/backends/imgui_impl_opengl3.cpp
+++ b/backends/imgui_impl_opengl3.cpp
@@ -99,13 +99,7 @@
 
 // GL includes
 #if defined(IMGUI_IMPL_OPENGL_ES2)
-#include <GLES2/gl2.h>
-#if defined(__EMSCRIPTEN__)
-#ifndef GL_GLEXT_PROTOTYPES
-#define GL_GLEXT_PROTOTYPES
-#endif
-#include <GLES2/gl2ext.h>
-#endif
+#include <gles2_inc.h>
 #elif defined(IMGUI_IMPL_OPENGL_ES3)
 #if defined(__APPLE__)
 #include <TargetConditionals.h>
-- 
2.33.0

