diff --git a/src/client/main.cpp b/src/client/main.cpp
index 0aad12f..591158d 100644
--- a/src/client/main.cpp
+++ b/src/client/main.cpp
@@ -136,6 +136,7 @@ void DrawpileApp::openUrl(QUrl url)
 QStringList DrawpileApp::dataPaths()
 {
 	QStringList datapaths;
+	datapaths << "share/drawpile/drawpile";
 #ifndef Q_OS_MAC
 	datapaths << qApp->applicationDirPath();
 	datapaths << QStandardPaths::standardLocations(QStandardPaths::DataLocation);
