diff -ru kdnssd-5.54.0.orig/src/mdnsd-publicservice.cpp kdnssd-5.54.0/src/mdnsd-publicservice.cpp
--- kdnssd-5.54.0.orig/src/mdnsd-publicservice.cpp	2019-01-20 17:23:54.407549801 +0000
+++ kdnssd-5.54.0/src/mdnsd-publicservice.cpp	2019-01-20 17:24:50.534010980 +0000
@@ -20,7 +20,7 @@
 
 #include <QCoreApplication>
 #include <QStringList>
-#include <netinet/in.h>
+#include <QtEndian>
 #include "publicservice.h"
 #include "servicebase_p.h"
 #include "mdnsd-sdevent.h"
@@ -172,7 +172,7 @@
         fullType += ',' + subtype;
     }
     if (DNSServiceRegister(&ref, 0, 0, d->m_serviceName.toUtf8().constData(), fullType.toLatin1().constData(), domainToDNS(d->m_domain).constData(), NULL,
-                           htons(d->m_port), TXTRecordGetLength(&txt), TXTRecordGetBytesPtr(&txt), publish_callback,
+                           qToBigEndian(d->m_port), TXTRecordGetLength(&txt), TXTRecordGetBytesPtr(&txt), publish_callback,
                            reinterpret_cast<void *>(d)) == kDNSServiceErr_NoError) {
         d->setRef(ref);
     }
diff -ru kdnssd-5.54.0.orig/src/mdnsd-remoteservice.cpp kdnssd-5.54.0/src/mdnsd-remoteservice.cpp
--- kdnssd-5.54.0.orig/src/mdnsd-remoteservice.cpp	2019-01-20 17:23:54.407549801 +0000
+++ kdnssd-5.54.0/src/mdnsd-remoteservice.cpp	2019-01-20 17:24:32.840742420 +0000
@@ -18,7 +18,7 @@
  * Boston, MA 02110-1301, USA.
  */
 
-#include <netinet/in.h>
+#include <QtEndian>
 #include <QEventLoop>
 #include <QCoreApplication>
 #include <QDebug>
@@ -152,7 +152,7 @@
             map[QString::fromUtf8(key)].clear();
         }
     }
-    ResolveEvent rev(DNSToDomain(hosttarget), ntohs(port), map);
+    ResolveEvent rev(DNSToDomain(hosttarget), qFromBigEndian(port), map);
     QCoreApplication::sendEvent(obj, &rev);
 }
 
